pkg/agent/controller/l7visibility/l7visibility_controller.go

```diff
@@ -0,0 +1,412 @@
+	if oldAnnotation == updatedAnnotation {
+		return
+	}
+
+	klog.V(2).InfoS("Processing Namespace UPDATE event", "Namespace", klog.KObj(updatedNS))
+
+	affectedPods := l7c.getNonAnnotatedPodsFromNS(updatedNS)
+	for _, pod := range affectedPods {
+		podNN := k8s.NamespacedName(pod.Namespace, pod.Name)
+		l7c.queue.Add(podNN)
+	}
+}
+
+func (l7c *L7VisibilityController) getNonAnnotatedPodsFromNS(ns *v1.Namespace) []*v1.Pod {
+	var nonAnnotatedPods []*v1.Pod
+	pods, _ := l7c.podLister.Pods(ns.Name).List(labels.Everything())
+
+	// Only select the non annotated Pods, as annotated Pods are handled separately
+	for _, pod := range pods {
+		_, ok := pod.Annotations[types.L7VisibilityAnnotationKey]
+		if !ok {
+			nonAnnotatedPods = append(nonAnnotatedPods, pod)
+		}
+	}
+	return nonAnnotatedPods
+}
+
+func (l7c *L7VisibilityController) syncPod(podNN string) error {
+	podNamespace, podName := k8s.SplitNamespacedName(podNN)
+	pod, err := l7c.podLister.Pods(podNamespace).Get(podName)
+	if err != nil {
+		// Remove the TC flows if the Pod has been deleted
+		return l7c.removeTCFlow(podNN)
+	}
+	podNS, err := l7c.namespaceLister.Get(pod.Namespace)
+	if err != nil {
+		// Remove TC flows if Namespace has been deleted
+		return l7c.removeTCFlow(podNN)
+	}
+
+	var annotationValue string
+	podAnnotationValue, ok := pod.Annotations[types.L7VisibilityAnnotationKey]
+	if !ok {
+		// Both Pod and Namespace are not annotated, remove the TC Mark flow
+		nsAnnotationValue, ok := podNS.Annotations[types.L7VisibilityAnnotationKey]
+		if !ok {
+			return l7c.removeTCFlow(podNN)
+		}
+		annotationValue = nsAnnotationValue
+	}
+	// Pod has annotation, we will add it to the TCmarkflow
+	if annotationValue == "" {
+		annotationValue = podAnnotationValue
+	}
+
+	// Check if the annotation value is one of the specified values
+	var direction v1alpha2.Direction
+	annotationValue = strings.ToLower(annotationValue)
+	switch annotationValue {
+	case "ingress":
+		direction = v1alpha2.DirectionIngress
+	case "egress":
+		direction = v1alpha2.DirectionEgress
+	case "both":
+		direction = v1alpha2.DirectionBoth
+	default:
+		return errInvalidAnnotation
+	}
+	podInterfaces := l7c.interfaceStore.GetContainerInterfacesByPod(pod.Name, pod.Namespace)
+	if len(podInterfaces) == 0 {
+		return errPodInterfaceNotFound
+	}
+	sourceOfPort := []uint32{uint32(podInterfaces[0].OFPort)}
+
+	// Start Suricata before starting traffic control mark flows
+	l7c.l7Reconciler.StartSuricataOnce()
+
+	// Return if Pod with direction is already present in map
+	if l7c.ifPodNNAndDirectionExistsInMap(podNN, direction) {
+		return nil
+	}
+	// Remove TC mark flows if PodNN is present in map with different direction
+	l7c.removeTCFlow(podNN)
+
+	tcName := fmt.Sprintf("tcl7:%s", podNN)
+	if err := l7c.ofClient.InstallTrafficControlMarkFlows(tcName, sourceOfPort, l7c.targetPort, direction, v1alpha2.ActionMirror); err != nil {
+		return err
+	}
+	l7c.addPodNNDirToMap(podNN, direction)
```

Comment:
The functions names are uncommon "ifXXXXX" and a bit hard to get their meaning, and error handling is missing. I would suggest the following code:
```suggestion
	oldDirection, exists := l7c.getMirroredDirection(podNN)
	if exists {
		if oldDirection == direction {
			return nil
		}
		if err := l7c.removeTCFlow(podNN); err != nil {
			return err
		}
	}
	tcName := fmt.Sprintf("tcl7:%s", podNN)
	if err := l7c.ofClient.InstallTrafficControlMarkFlows(tcName, sourceOfPort, l7c.targetPort, direction, v1alpha2.ActionMirror); err != nil {
		return err
	}
	l7c.updateMirroredDirection(podNN, direction)
```

Reply:
not addressed